src/graphnet/models/components/layers.py

```diff
@@ -593,3 +602,393 @@ def forward(
+                requires_grad=True,
+            )
+            nn.init.xavier_normal_(self.VeRow)
+
+    def forward(self, data: Data) -> Data:
+        """Forward pass."""
+        Q_x = self.Q(data.x)
+        K_x = self.K(data.x)
+        V_x = self.V(data.x)
+
+        if data.get("edge_attr", None) is not None:
+            E = self.E(data.edge_attr)
+        else:
+            E = None
+
+        Q_x = Q_x.view(-1, self.num_heads, self.out_dim)
+        K_x = K_x.view(-1, self.num_heads, self.out_dim)
+        V_x = V_x.view(-1, self.num_heads, self.out_dim)
+
+        # Applying Eq. 2 of the GRIT paper:
+        src = K_x[data.edge_index[0]]  # (num relative) x num_heads x out_dim
+        dest = Q_x[data.edge_index[1]]  # (num relative) x num_heads x out_dim
+        score = src + dest  # element-wise multiplication
+        if E is not None:
+            E = E.view(-1, self.num_heads, self.out_dim * 2)
+            E_w, E_b = E[:, :, : self.out_dim], E[:, :, self.out_dim :]
+            score = score * E_w
+            score = torch.sqrt(torch.relu(score)) - torch.sqrt(
+                torch.relu(-score)
+            )
+            score = score + E_b
+
+        score = self.activation(score)
+        e_t = score  # ehat_ij
+
+        # Output edge
+        if E is not None:
+            wE = score.flatten(1)
+
+        # Complete attention ccaclculation
+        score = torch.einsum("ehd, dhc->ehc", score, self.Aw)
+        if self.clamp is not None:
+            score = torch.clamp(score, min=-self.clamp, max=self.clamp)
+        score = softmax(score, index=data.edge_index[1]).to(
+            dtype=data.x.dtype
+        )  # (num relative) x num_heads x 1
+        score = self.dropout(score)
+
+        # Aggregate with Attn-Score
+        V_x_weighted = (
+            V_x[data.edge_index[0]] * score
+        )  # (num relative) x num_heads x out_dim
+        wV = torch.zeros_like(
+            V_x, dtype=score.dtype
+        )  # (num nodes in batch) x num_heads x out_dim
+        scatter(V_x_weighted, data.edge_index[1], dim=0, out=wV, reduce="add")
+
+        # Adds the second term (W_Ev ehhat_ij) in the last line of Eq. 2
+        if self.edge_enhance and E is not None:
+            rowV = scatter(
+                e_t * score, data.edge_index[1], dim=0, reduce="add"
+            )
+            rowV = torch.einsum("nhd, dhc -> nhc", rowV, self.VeRow)
+            wV = wV + rowV
+
+        return wV, wE
+
+
+class GritTransformerLayer(LightningModule):
+    """Proposed Transformer Layer for GRIT.
+
+    Original code:
+    https://github.com/LiamMa/GRIT/blob/main/grit/layer/grit_layer.py
+    """
+
+    def __init__(
+        self,
+        in_dim: int,
+        out_dim: int,
+        num_heads: int,
+        dropout: float = 0.0,
+        norm: nn.Module = nn.BatchNorm1d,
+        residual: bool = True,
+        deg_scaler: bool = True,
+        activation: nn.Module = nn.ReLU,
```

Comment:
Could mention here that the activation function is assumed to be un-instantiated. E.g.
```
"""
activation:  Reference to uninstantiated activation function. E.g. `torch.nn.LeakyReLU`
"""
```